### ISSUE TITLE
SubSurface Type reset when assigned to surface

### ISSUE THREAD

**Issue by shorowit** [NREL/OpenStudio#3936]
Issue overview
--------------
Running the following script:
```
require 'openstudio'

model = OpenStudio::Model::Model.new

surf_polygon = OpenStudio::Point3dVector.new
surf_polygon << OpenStudio::Point3d.new(0, 0, 0)
surf_polygon << OpenStudio::Point3d.new(0, 10, 0)
surf_polygon << OpenStudio::Point3d.new(0, 10, 10)
surf_polygon << OpenStudio::Point3d.new(0, 0, 10)
surf = OpenStudio::Model::Surface.new(surf_polygon, model)

door_polygon = OpenStudio::Point3dVector.new
door_polygon << OpenStudio::Point3d.new(0, 1, 1)
door_polygon << OpenStudio::Point3d.new(0, 9, 1)
door_polygon << OpenStudio::Point3d.new(0, 9, 9)
door_polygon << OpenStudio::Point3d.new(0, 1, 9)
subsurf = OpenStudio::Model::SubSurface.new(door_polygon, model)
subsurf.setSubSurfaceType('Door')

puts subsurf.subSurfaceType # should be Door, and is Door

subsurf.setSurface(surf)

puts subsurf.subSurfaceType # should still be Door, but is now FixedWindow
```

I expect it to say Door twice, but instead it produces:
```
Door
FixedWindow
```

I can workaround this by setting the subsurface type _after_ the door is assigned to the surface.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Windows 10
 - Version of OpenStudio (if using an intermediate build, include SHA): OpenStudio 2.9.1

**Comment by joseph-robertson** [comment]
Doesn't the door subsurface need to be adjacent to the bottom of the surface? Seems like in your example above, the door sits slightly above the bottom of the surface.

**Comment by shorowit** [comment]
I don't think it should have to be. I told OpenStudio it's a door, it shouldn't override it.

It's true that if you change the subsurface vertices so that the door is located on the bottom edge, you get the correct results. But I'm assuming you get the correct results for the wrong reason, which is that OS again overrode my user-entered value but happened to default to Door this time when it's along the bottom edge.

**Comment by jmarrec** [comment]
It's working as intended right now: OpenStudio doesn't know whether you told it it's a door or if it just is the constructor's default. This is a required field, it's assigned in the Ctor by figuring out the default https://github.com/NREL/OpenStudio/blob/0cb319a4977599dbe3eca55fc39abdfa077aeaa7/src/model/SubSurface.cpp#L1158-L1163

`setSurfaceType` will call `assignDefaultSubSurfaceType` if it wasn't previously already on a `Surface`.

https://github.com/NREL/OpenStudio/blob/0cb319a4977599dbe3eca55fc39abdfa077aeaa7/src/model/SubSurface.cpp#L674-L682

I think this is right, and drawing geometry in sketchup would be extremely awkward if it didn't do that. If you want to specifically go around that, just tell it it's a door AFTER you put it on a surface.

**Comment by jmarrec** [comment]
If you really don't like it @shorowit I suppose you can use monkey-patching, for example using Module#prepend, and understanding that this will not distinguish between cases where you explicitly called `setSubSurfaceType` previously or not.

```ruby
module SubSurfaceExtensions
  def setSurface(surf)
    old_type = self.subSurfaceType
    super(surf)
    self.setSubSurfaceType(old_type)
  end
end

class OpenStudio::Model::SubSurface
  prepend SubSurfaceExtensions
end

model = OpenStudio::Model::Model.new

surf_polygon = OpenStudio::Point3dVector.new
surf_polygon << OpenStudio::Point3d.new(0, 0, 0)
surf_polygon << OpenStudio::Point3d.new(0, 10, 0)
surf_polygon << OpenStudio::Point3d.new(0, 10, 10)
surf_polygon << OpenStudio::Point3d.new(0, 0, 10)
surf = OpenStudio::Model::Surface.new(surf_polygon, model)

door_polygon = OpenStudio::Point3dVector.new
door_polygon << OpenStudio::Point3d.new(0, 1, 1)
door_polygon << OpenStudio::Point3d.new(0, 9, 1)
door_polygon << OpenStudio::Point3d.new(0, 9, 9)
door_polygon << OpenStudio::Point3d.new(0, 1, 9)
subsurf = OpenStudio::Model::SubSurface.new(door_polygon, model)
subsurf.setSubSurfaceType('Door')

puts subsurf.subSurfaceType # should be Door, and is Door

subsurf.setSurface(surf)

puts subsurf.subSurfaceType # Still Door, since I have monkeypatched it
```

**Comment by shorowit** [comment]
I see. Isn't this issue handled in other objects by having an IsFooDefaulted method/property and keeping track of whether the value is being defaulted? For example, [ThermalZone::isMultiplierDefaulted](https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-2.9.0-doc/model/html/classopenstudio_1_1model_1_1_thermal_zone.html).

**Comment by jmarrec** [comment]
I mentioned that in my last reply above. I think the difference is that sketchup would often (potentially all the time...?) have to query it to know it needs to render it as a window (blue, transparent) or a door (brown, opaque), and that's why I said I think it'd be very slow, not 100% sure.

**Comment by shorowit** [comment]
Ah sorry, Julien. I read your previous response and before you had edited it, so I didn't see where you mentioned the IsFooDefaulted alternative.

Just to be clear, I never said this was an urgent or high priority need. And I mentioned in the original issue description that we are now using a workaround. But it sure seems like a bug to me when the SDK overrides a property value you set simply because you assigned the subsurface to a parent surface. No one would expect this behavior. We had a bug in our code because of this for many years without realizing it.

I can't really imagine why Sketchup the isFooDefaulted method would affect SketchUp, I would have thought the logic lives in the SDK and SketchUp wouldn't have to change a thing. But I'm not familiar with how SketchUp works, so I'll take your word for it.

Long story short, I believe we should just keep this issue open as a low priority bug.

**Comment by tijcolem** [comment]
@kbenne Maybe we can start slack channel to discuss in more detail and get group consensus?

**Comment by joseph-robertson** [comment]
@shorowit @jmarrec 

Let's say I:
* changed the idd from required-field to default (FixedWindow) for SubSurface
* removed assignDefaultSubSurfaceType from the ctor for SubSurface
* then in setSurface only called assignDefaultSubSurfaceType if the SubSurfaceType field is empty

Seems like the previous would solve the issue of explicitly calling the subsurface a Door, calling setSurface, and then having the Door change to, e.g., FixedWindow. But what about the other way? Wouldn't you have a problem if you explicitly called setSubSurfaceType(FixedWindow) (field is blank), then called setSurface, but the subsurface's vertices implied it was, e,g., a door?
